c++/src/kj/common.h

```diff
@@ -1910,22 +1910,18 @@ class ArrayPtr: public DisallowConstCopyIfNotConst<T> {
     return { reinterpret_cast<PropagateConst<T, char>*>(ptr), size_ * sizeof(T) };
   }
 
-  inline bool operator==(decltype(nullptr)) const { return size_ == 0; }
+  inline constexpr bool operator==(decltype(nullptr)) const { return size_ == 0; }
 
-  inline bool operator==(const ArrayPtr& other) const {
+  inline constexpr bool operator==(const ArrayPtr& other) const {
     if (size_ != other.size_) return false;
-    if (isIntegral<RemoveConst<T>>()) {
-      if (size_ == 0) return true;
-      return memcmp(ptr, other.ptr, size_ * sizeof(T)) == 0;
```

Comment:
Removing the memcmp() optimization here might have negative consequences. I think @mikea investigated this carefully fairly recently and found that the compiler doesn't always figure out that the for loop is equivalent to memcmp.

Is there some way we can tell the compiler to use a different codepath when evaluating code for constexpr purposes vs. runtime? Or maybe we can use __builtin_memcmp() which someone told me is constexpr?

Reply:
`__builtin_memcmp` is indeed `constexpr`.